File: code/packages/go/multi-directed-graph/multi_directed_graph_test.go
package multidirectedgraph

import (
	"errors"
	"math"
	"reflect"
	"testing"
)

func TestKeepsParallelEdgesWithStableIDsAndProperties(t *testing.T) {
	graph := New[string]()
	first, err := graph.AddEdge("a", "b", 0.25, nil, "")
	if err != nil {
		t.Fatal(err)
	}
	second, err := graph.AddEdge("a", "b", 0.75, PropertyBag{"trainable": true}, "w1")
	if err != nil {
		t.Fatal(err)
	}

	if first != "e0" || second != "w1" {
		t.Fatalf("unexpected edge IDs: %s %s", first, second)
	}
	edges, err := graph.EdgesBetween("a", "b")
	if err != nil {
		t.Fatal(err)
	}
	if len(edges) != 2 {
		t.Fatalf("parallel edge count = %d, want 2", len(edges))
	}
	successors, err := graph.Successors("a")
	if err != nil {
		t.Fatal(err)
	}
	predecessors, err := graph.Predecessors("b")
	if err != nil {
		t.Fatal(err)
	}
	if !reflect.DeepEqual(successors, []string{"b"}) || !reflect.DeepEqual(predecessors, []string{"a"}) {
		t.Fatalf("parallel edges were not deduplicated: successors=%v predecessors=%v", successors, predecessors)
	}
	properties, err := graph.EdgeProperties("w1")
	if err != nil || properties["trainable"] != true || properties["weight"] != 0.75 {
		t.Fatalf("edge properties not preserved: properties=%#v err=%v", properties, err)
	}
}

func TestAutoAllocatedEdgeIDsAvoidExplicitIDs(t *testing.T) {
	graph := New[string]()
	if _, err := graph.AddEdge("a", "b", 1, nil, "e0"); err != nil {
		t.Fatal(err)
	}
	edgeID, err := graph.AddEdge("b", "c", 1, nil, "")
	if err != nil {
		t.Fatal(err)
	}
	if edgeID != "e1" {
		t.Fatalf("edge ID = %q, want e1", edgeID)
	}
}

func TestRejectsDuplicateEdgeIDs(t *testing.T) {
	graph := New[string]()
	if _, err := graph.AddEdge("a", "b", 1, nil, "edge"); err != nil {
		t.Fatal(err)
	}
	_, err := graph.AddEdge("a", "c", 1, nil, "edge")
	var duplicate DuplicateEdgeIDError
	if !errors.As(err, &duplicate) || duplicate.EdgeID != "edge" {
		t.Fatalf("expected DuplicateEdgeIDError for edge, got %v", err)
	}
}

func TestCopiesAndMutatesPropertyBags(t *testing.T) {
	graph := New[string]()
	nodeInput := PropertyBag{"role": "input"}
	graph.AddNode("a", nodeInput)
	graph.AddNode("a", PropertyBag{"shape": "[1]"})
	nodeInput["role"] = "mutated"

	graph.SetGraphProperty("name", "generic")
	graph.SetGraphProperty("version", 1)
	graph.RemoveGraphProperty("version")
	graphProperties := graph.GraphProperties()
	graphProperties["name"] = "mutated"
	if graph.GraphProperties()["name"] != "generic" {
		t.Fatal("graph properties leaked a mutable internal map")
	}

	nodeProperties, err := graph.NodeProperties("a")
	if err != nil {
		t.Fatal(err)
	}
	if nodeProperties["role"] != "input" || nodeProperties["shape"] != "[1]" {
		t.Fatalf("node properties = %#v", nodeProperties)
	}
	nodeProperties["role"] = "mutated"
	if err := graph.SetNodeProperty("a", "kind", "feature"); err != nil {
		t.Fatal(err)
	}
	if err := graph.RemoveNodeProperty("a", "shape"); err != nil {
		t.Fatal(err)
	}
	nodeProperties, err = graph.NodeProperties("a")
	if err != nil {
		t.Fatal(err)
	}
	if nodeProperties["role"] != "input" || nodeProperties["kind"] != "feature" {
		t.Fatalf("node property updates = %#v", nodeProperties)
	}
	if _, exists := nodeProperties["shape"]; exists {
		t.Fatalf("removed node property still present: %#v", nodeProperties)
	}
}

func TestSynchronizesWeightProperty(t *testing.T) {
	graph := New[string]()
	input := PropertyBag{"channel": "left"}
	if _, err := graph.AddEdge("a", "b", 2, input, "w"); err != nil {
		t.Fatal(err)
	}
	input["channel"] = "mutated"
	properties, err := graph.EdgeProperties("w")
	if err != nil || properties["channel"] != "left" {
		t.Fatalf("edge property input was not copied: properties=%#v err=%v", properties, err)
	}
	properties["channel"] = "mutated"
	properties, _ = graph.EdgeProperties("w")
	if properties["channel"] != "left" {
		t.Fatal("edge properties leaked a mutable internal map")
	}

	if err := graph.SetEdgeProperty("w", "weight", int32(3)); err != nil {
		t.Fatal(err)
	}
	weight, err := graph.EdgeWeight("w")
	if err != nil || weight != 3 {
		t.Fatalf("weight = %v err=%v, want 3", weight, err)
	}
	if err := graph.SetEdgeProperty("w", "weight", "heavy"); err == nil {
		t.Fatal("expected a non-numeric weight error")
	}
	if err := graph.SetEdgeProperty("w", "weight", math.NaN()); err == nil {
		t.Fatal("expected a NaN weight error")
	}
	if err := graph.RemoveEdgeProperty("w", "weight"); err != nil {
		t.Fatal(err)
	}
	weight, err = graph.EdgeWeight("w")
	if err != nil || weight != 1 {
		t.Fatalf("weight after removing property = %v err=%v, want 1", weight, err)
	}
}

func TestControlsSelfLoops(t *testing.T) {
	graph := New[string]()
	if _, err := graph.AddEdge("a", "a", 1, nil, ""); err == nil {
		t.Fatal("expected self-loop rejection")
	}

	graph = NewAllowSelfLoops[string]()
	edgeID, err := graph.AddEdge("a", "a", 1, nil, "")
	if err != nil || !graph.HasEdge(edgeID) || !graph.HasCycle() {
		t.Fatalf("allowed self-loop was not retained: edge=%q err=%v", edgeID, err)
	}
}

func TestMissingNodesAndEdgesReturnTypedErrors(t *testing.T) {
	graph := New[string]()

	nodeChecks := []func() error{
		func() error { _, err := graph.NodeProperties("missing"); return err },
		func() error { _, err := graph.OutgoingEdges("missing"); return err },
		func() error { _, err := graph.IncomingEdges("missing"); return err },
		func() error { _, err := graph.Successors("missing"); return err },
		func() error { _, err := graph.Predecessors("missing"); return err },
		func() error { return graph.SetNodeProperty("missing", "key", "value") },
		func() error { return graph.RemoveNodeProperty("missing", "key") },
		func() error { return graph.RemoveNode("missing") },
	}
	for index, check := range nodeChecks {
		var target NodeNotFoundError[string]
		if err := check(); !errors.As(err, &target) {
			t.Fatalf("node check %d returned %v, want NodeNotFoundError", index, err)
		}
	}

	edgeChecks := []func() error{
		func() error { _, err := graph.Edge("missing"); return err },
		func() error { _, err := graph.EdgeWeight("missing"); return err },
		func() error { _, err := graph.EdgeProperties("missing"); return err },
		func() error { return graph.SetEdgeProperty("missing", "key", "value") },
		func() error { return graph.RemoveEdgeProperty("missing", "key") },
		func() error { return graph.RemoveEdge("missing") },
	}
	for index, check := range edgeChecks {
		var target EdgeNotFoundError
		if err := check(); !errors.As(err, &target) {
			t.Fatalf("edge check %d returned %v, want EdgeNotFoundError", index, err)
		}
	}
}

func TestRemovesEdgesAndIncidentEdges(t *testing.T) {
	graph := New[string]()
	incoming, _ := graph.AddEdge("a", "b", 1, nil, "")
	outgoing, _ := graph.AddEdge("b", "c", 1, nil, "")
	kept, _ := graph.AddEdge("a", "c", 1, nil, "")

	if err := graph.RemoveEdge(kept); err != nil {
		t.Fatal(err)
	}
	if graph.HasEdge(kept) {
		t.Fatal("removed edge still exists")
	}
	if err := graph.RemoveNode("b"); err != nil {
		t.Fatal(err)
	}
	if graph.HasNode("b") || graph.HasEdge(incoming) || graph.HasEdge(outgoing) {
		t.Fatalf("node or incident edges remain: nodes=%v edges=%v", graph.Nodes(), graph.Edges())
	}
}

func TestTopologicalSortAndIndependentGroupsAccountForParallelEdges(t *testing.T) {
	graph := New[string]()
	graph.AddEdge("a", "c", 1, nil, "")
	graph.AddEdge("a", "c", 1, nil, "")
	graph.AddEdge("b", "c", 1, nil, "")
	graph.AddEdge("c", "d", 1, nil, "")

	order, err := graph.TopologicalSort()
	if err != nil {
		t.Fatal(err)
	}
	if got := join(order); got != "a,b,c,d" {
		t.Fatalf("order = %s", got)
	}
	groups, err := graph.IndependentGroups()
	if err != nil {
		t.Fatal(err)
	}
	if join(groups[0]) != "a,b" || join(groups[1]) != "c" || join(groups[2]) != "d" {
		t.Fatalf("groups = %#v", groups)
	}
}

func TestCycleErrorsIncludeProgress(t *testing.T) {
	graph := New[string]()
	graph.AddNode("ready", nil)
	graph.AddEdge("a", "b", 1, nil, "")
	graph.AddEdge("b", "a", 1, nil, "")

	_, err := graph.TopologicalSort()
	var cycle CycleError
	if !errors.As(err, &cycle) || cycle.Processed != 1 || cycle.Total != 3 {
		t.Fatalf("topological cycle error = %#v", err)
	}
	_, err = graph.IndependentGroups()
	if !errors.As(err, &cycle) || !graph.HasCycle() {
		t.Fatalf("independent groups cycle error = %#v", err)
	}
}

func TestSupportsGenericNodeValues(t *testing.T) {
	graph := New[int]()
	if graph.Size() != 0 {
		t.Fatalf("new graph size = %d", graph.Size())
	}
	graph.AddEdge(2, 3, 1, nil, "")
	graph.AddNode(1, nil)

	order, err := graph.TopologicalSort()
	if err != nil {
		t.Fatal(err)
	}
	if !reflect.DeepEqual(order, []int{1, 2, 3}) {
		t.Fatalf("generic node order = %v", order)
	}
}

func join(values []string) string {
	result := ""
	for index, value := range values {
		if index > 0 {
			result += ","
		}
		result += value
	}
	return result
}

File: code/packages/typescript/multi-directed-graph/tests/multi-directed-graph.test.ts
import { describe, expect, it } from "vitest";

import {
  DuplicateEdgeIdError,
  EdgeNotFoundError,
  MultiDirectedGraph,
  MultiDirectedGraphCycleError,
  NodeNotFoundError,
} from "../src/index.js";

describe("MultiDirectedGraph", () => {
  it("adds nodes and merges node metadata", () => {
    const graph = new MultiDirectedGraph();

    graph.addNode("input", { "nn.op": "input" });
    graph.addNode("input", { "nn.shape": "[1]" });

    expect(graph.nodes()).toEqual(["input"]);
    expect(graph.nodeProperties("input")).toEqual({
      "nn.op": "input",
      "nn.shape": "[1]",
    });
  });

  it("stores parallel directed edges with stable IDs", () => {
    const graph = new MultiDirectedGraph();

    const e0 = graph.addEdge("A", "B", 0.25, { channel: "left" }, "edge-a");
    const e1 = graph.addEdge("A", "B", 0.75, { channel: "right" }, "edge-b");

    expect(e0).toBe("edge-a");
    expect(e1).toBe("edge-b");
    expect(graph.edgesBetween("A", "B").map((edge) => edge.id)).toEqual([
      "edge-a",
      "edge-b",
    ]);
    expect(graph.successors("A")).toEqual(["B"]);
    expect(graph.predecessors("B")).toEqual(["A"]);
  });

  it("supports generic node values", () => {
    const graph = new MultiDirectedGraph<number>();

    const edgeId = graph.addEdge(1, 2, 3.5);

    expect(graph.nodes()).toEqual([1, 2]);
    expect(graph.edge(edgeId)).toEqual({
      id: edgeId,
      from: 1,
      to: 2,
      weight: 3.5,
    });
    expect(graph.successors(1)).toEqual([2]);
    expect(graph.predecessors(2)).toEqual([1]);
  });

  it("keeps edge properties and weights synchronized", () => {
    const graph = new MultiDirectedGraph();

    const edgeId = graph.addEdge("x0", "sum", 0.5, {
      "nn.trainable": true,
    });

    expect(graph.edgeProperties(edgeId)).toEqual({
      "nn.trainable": true,
      weight: 0.5,
    });

    graph.setEdgeProperty(edgeId, "weight", 0.75);
    expect(graph.edgeWeight(edgeId)).toBe(0.75);
    expect(graph.edgeProperties(edgeId).weight).toBe(0.75);

    graph.removeEdgeProperty(edgeId, "weight");
    expect(graph.edgeWeight(edgeId)).toBe(1.0);
    expect(graph.edgeProperties(edgeId).weight).toBe(1.0);
  });

  it("stores graph properties", () => {
    const graph = new MultiDirectedGraph();

    graph.setGraphProperty("nn.name", "tiny-model");
    graph.setGraphProperty("nn.version", "0");
    expect(graph.graphProperties()).toEqual({
      "nn.name": "tiny-model",
      "nn.version": "0",
    });

    graph.removeGraphProperty("nn.version");
    expect(graph.graphProperties()).toEqual({ "nn.name": "tiny-model" });
  });

  it("removes incident edges and metadata when removing a node", () => {
    const graph = new MultiDirectedGraph();

    const e0 = graph.addEdge("A", "B", 1.0, { role: "out" });
    const e1 = graph.addEdge("C", "A", 2.0, { role: "in" });

    graph.removeNode("A");

    expect(graph.hasNode("A")).toBe(false);
    expect(graph.hasEdge(e0)).toBe(false);
    expect(graph.hasEdge(e1)).toBe(false);
    expect(() => graph.edgeProperties(e0)).toThrow(EdgeNotFoundError);
    expect(() => graph.nodeProperties("A")).toThrow(NodeNotFoundError);
  });

  it("topologically sorts with parallel edge counts", () => {
    const graph = new MultiDirectedGraph();

    graph.addEdge("A", "B");
    graph.addEdge("A", "B");
    graph.addEdge("B", "C");

    expect(graph.topologicalSort()).toEqual(["A", "B", "C"]);
    expect(graph.independentGroups()).toEqual([["A"], ["B"], ["C"]]);
    expect(graph.hasCycle()).toBe(false);
  });

  it("detects cycles", () => {
    const graph = new MultiDirectedGraph();

    graph.addEdge("A", "B");
    graph.addEdge("B", "A");

    expect(graph.hasCycle()).toBe(true);
    expect(() => graph.topologicalSort()).toThrow(MultiDirectedGraphCycleError);
  });

  it("controls self-loops", () => {
    expect(() => new MultiDirectedGraph().addEdge("A", "A")).toThrow(
      "Self-loops are not allowed"
    );

    const graph = new MultiDirectedGraph({ allowSelfLoops: true });
    const edgeId = graph.addEdge("A", "A");
    expect(graph.edge(edgeId)).toEqual({
      id: edgeId,
      from: "A",
      to: "A",
      weight: 1,
    });
  });

  it("rejects duplicate edge IDs", () => {
    const graph = new MultiDirectedGraph();

    graph.addEdge("A", "B", 1.0, {}, "edge");

    expect(() => graph.addEdge("A", "B", 1.0, {}, "edge")).toThrow(
      DuplicateEdgeIdError
    );
  });

  it("returns useful string summaries", () => {
    const graph = new MultiDirectedGraph();
    graph.addEdge("A", "B");

    expect(graph.toString()).toBe("MultiDirectedGraph(nodes=2, edges=1)");
  });
});
